goserver/session/session.go

```diff
@@ -0,0 +1,114 @@
+package session
+
+import (
+	"crypto/hmac"
+	"crypto/sha1"
+	"encoding/base64"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+)
+
+var (
+	// SessionCookieSecret is used to encrypt the cookie session id.
+	SessionCookieSecret = ""
+)
+
+const (
+	sessionLifeInSec = 7200
+	cookieLifeInSec  = 7200
+	cookieName       = "session"
+)
+
+// Session stores necessary information for a telehealth session.
+type Session struct {
+	ID        string
+	ExpiresAt time.Time
+	Value     map[string]interface{}
+}
+
+// Put stores the key-val pair in Session.
+func (s *Session) Put(key string, val interface{}) {
+	if s.Value == nil {
+		s.Value = make(map[string]interface{})
+	}
+	s.Value[key] = val
+}
+
+// Get returns the value for the given key.
+func (s *Session) Get(key string) interface{} {
+	if s.Value == nil {
+		return nil
+	}
+	v, ok := s.Value[key]
+	if !ok {
+		return nil
+	}
+	return v
+}
+
+// New creates a new session and set cookie containning the encoded session id.
+func New(m *StoreManager, w http.ResponseWriter, r *http.Request) (*Session, error) {
+	expireAt := time.Now().Add(sessionLifeInSec * time.Second)
+	s, err := m.Create(expireAt)
+	if err != nil {
+		return nil, err
+	}
+	expiration := time.Now().Add(cookieLifeInSec * time.Second)
+	cookie := &http.Cookie{Name: cookieName, Value: encodeSessionID(s.ID), Expires: expiration}
+	http.SetCookie(w, cookie)
+	r.AddCookie(cookie)
```

Comment:
we probably don't need to set cookie on request?

Reply:
This is a standard way how session cookie should be handled (e.g., http://expressjs.com/en/resources/middleware/cookie-session.html)- the cookie should be set in the current request as well so that things like middlewares can be introduced.

Reply:
I see. Would be nice to add a comment regarding middleware.